Add implementation for rfc-1 (files)

rylev opened the pull request:
Implementation for [rfc-1](https://github.com/rustwasm/gloo/blob/master/rfcs/001-mid-level-file-api.md).

This implementation is not complete. Left to do:
* [ ] Finalize design for errors
* [ ] Implement event based callbacks
* [ ] Tests
* [ ] Documentation

fitzgen commented on crates/file/src/file_list.rs:
```diff
@@ -0,0 +1,23 @@
+    inner: Vec<File>,
+}
+
+impl std::convert::Into<FileList> for web_sys::FileList {
```
We should implement `From<web_sys::FileList> for FileList` and then we get the `Into` impl for free as well because there is a blanket `impl<T, U> Into<T> for U where T: From<U>`.

fitzgen replied:
Same for all the other `Into` impls. Generally, you never want to implement `Into` manually, only `From`.

Also, `Into` and `From` are in the prelude, so you don't need to fully qualify them.

rylev replied:
I always get this backwards. Will make the change.

Pauan commented on crates/file/src/file_reader.rs:
```diff
@@ -0,0 +1,248 @@
+            }
+        }
+
+        pub fn read_as_string(self, blob: &impl BlobLike) -> ReadAsString {
```
I think these should be top-level functions too.

Also, there's a lot of code duplication here. These should be using the `callbacks` functions and just wrapping them in Futures.

That means the `callbacks` functions need to return something which enables them to be dropped/cancelled.

Pauan commented on crates/file/src/file_reader.rs:
```diff
@@ -0,0 +1,248 @@
+                inner: web_sys::FileReader,
+            }
+
+            impl Future for $name {
```
I think we should seriously consider using Futures 0.3 instead.

The `Future` trait is already stabilized, and Futures 0.3 will be available on stable soon.

I'm also okay with delaying the `futures` submodule until after Futures 0.3 is available on stable (only releasing the `callbacks` submodule).

Pauan commented on crates/file/src/file_reader.rs:
```diff
@@ -0,0 +1,248 @@
+    AbortedEarly,
+    JsError(wasm_bindgen::JsValue),
+}
+impl std::convert::Into<FileReadError> for wasm_bindgen::JsValue {
```
This feels extremely broad. Since there are a limited number of errors which can be thrown, we shouldn't allow for converting anything into a `FileReadError`.

Pauan commented on crates/file/src/file_reader.rs:
```diff
@@ -0,0 +1,248 @@
+#[derive(Debug)]
+pub enum FileReadError {
+    AbortedEarly,
+    JsError(wasm_bindgen::JsValue),
```
This should take an `Error` object. Or even better, we can just list out all of the possible errors that can happen.

Pauan commented on crates/file/src/file_reader.rs:
```diff
@@ -0,0 +1,248 @@
+    }
+}
+
+impl std::convert::Into<FileReadError> for ::futures::sync::oneshot::Canceled {
```
I don't like this impl. It's exposing our internal implementation details.

Pauan commented on crates/file/src/file_reader.rs:
```diff
@@ -0,0 +1,248 @@
+
+impl std::fmt::Display for FileReadError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "error reading file")
```
This can probably be a lot nicer: for example printing the `Error` object's `message`.

Pauan commented on crates/file/src/file_reader.rs:
```diff
@@ -0,0 +1,248 @@
+            CallbackFn: FnOnce(Result<T, FileReadError>) + 'static,
+        {
+            let reader = self.inner.clone();
+            let closure = Closure::once(move || {
```
(I really should have caught this and the `self` issue during the RFC stage, sorry about that)

rylev commented on crates/file/src/blob.rs:
```diff
@@ -0,0 +1,200 @@
+        Blob::from_raw(inner.unwrap_throw())
+    }
+
+    pub fn from_raw(inner: web_sys::Blob) -> Blob {
```
This convention is mentioned in the contributing.md. Perhaps that should be changed?

Pauan replied:
I'm undecided. I guess `from_raw` is fine for now.

rylev commented on crates/file/src/blob.rs:
```diff
@@ -0,0 +1,200 @@
+            options.type_(&mime_type);
+        }
+        if let Some(last_modified) = last_modified_since_epoch {
+            let (millis, did_overflow_seconds) = last_modified.as_secs().overflowing_mul(1000);
```
Ah! I didn't realize `as_millis` had been stabilized! 🎉

rylev commented on crates/file/src/blob.rs:
```diff
@@ -0,0 +1,200 @@
+        File::from_raw(inner)
+    }
+
+    pub fn from_raw(inner: web_sys::File) -> File {
```
See reply above.

rylev commented:
@Pauan I believe I've addressed most of your issues. I've removed the futures based API for now since I think it complicates things and can easily be added in a following PR without breaking anything. Let me know what you think.

Of course, the work items I wrote out above still need to be addressed.

Pauan commented:
@rylev I'm in the middle of reviewing it, there's some tricky parts I need to think about.

fitzgen commented:
> @Pauan @fitzgen thoughts? Is this good enough to be cleaned up with tests and documentation?

In my opinion, definitely yes.

I think we are at the point where it makes more sense to get these docs and tests and merge this PR than it does to nitpick on too many little details right now. Once this is merged and a few other little things, we can cut a 0.1 release of gloo and we can start getting people using it and getting excitement around it and gathering feedback. I think those things would do a lot more for pushing this project forward and getting closer to fulfilling its goals.

Pauan commented on crates/file/src/file_reader.rs:
```diff
@@ -0,0 +1,110 @@
+    AbortedEarly,
+    JsError(js_sys::Error),
+}
+impl std::convert::From<wasm_bindgen::JsValue> for FileReadError {
```
This feels a bit odd: should it take `js_sys::Error` instead?

Pauan commented on crates/file/src/file_reader.rs:
```diff
@@ -0,0 +1,110 @@
+        });
+        let function = closure.as_ref().dyn_ref().unwrap_throw();
+
+        reader.clone().set_onload(Some(&function));
```
This `clone` doesn't seem necessary.

Pauan commented on crates/file/src/file_reader.rs:
```diff
@@ -0,0 +1,110 @@
+    use wasm_bindgen::{closure::Closure, JsCast, UnwrapThrowExt};
+
+    #[derive(Debug)]
+    pub struct FileRead {
```
I think it sounds a bit better if this is called `FileReader`

Pauan commented on crates/file/src/file_reader.rs:
```diff
@@ -0,0 +1,110 @@
+#[derive(Debug)]
+pub enum FileReadError {
+    AbortedEarly,
+    JsError(js_sys::Error),
```
I'm not 100% sure, but it sounds like [there's only 3 possible errors that can occur](https://w3c.github.io/FileAPI/#file-error-read), so maybe we want to list them out explicitly, rather than using a generic `Error` branch?

derekdreery commented:
@rylev would you mind if I picked this up and worked on it?

rylev commented:
@derekdreery I do not mind. I am unfortunately too busy to see this to the end. 😞

hamza1311 commented:
> This PR should be closed as it is superseeded by #83

This PR is #83...

derekdreery commented:
Ah yep apologies the PR I meant is #110
